### ISSUE TITLE
Compiler silently ignores associated const with trait type bounds

### ISSUE THREAD

**Issue by Robbepop** [rust-lang/rust#58547]
In the following code we have a trait `Y` with an associated constant `MY_CONST` that has a trait bound `X` instead of a concrete type which should be invalid in today's Rust as far as I understand the language.

However, the compiler does not complain at all.
Not about the actual definition in `Y` nor in the implementation of `Y for Foo` where `Bar` doesn't even match the required trait bound if that functionality was even implemented.

```rust
trait X {}
trait Y {
    const MY_CONST: X; // X is a trait! Compiler does not complain ...
}

struct Foo;
#[derive(Debug)]
struct Bar;

impl Foo {
    // We implement MY_CONST as Bar that does not implement X.
    // Again the compiler does not complain ...
    const MY_CONST: Bar = Bar;
}

fn main() {
    println!("{:?}", Foo::MY_CONST);
}
```

Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=80872b7e1b76e9ba65576137d097c125

**Comment by jonas-schievink** [comment]
The second impl is inherent, it doesn't implement `Y`, so the compiler won't complain there. If you change the code like this the error tells you what's going on:

```rust
trait X {}
trait Y {
    const MY_CONST: X; // X is a trait! Compiler does not complain ...
}

struct Foo;
#[derive(Debug)]
struct Bar;

impl Y for Foo {
    // We implement MY_CONST as Bar that does not implement X.
    // Again the compiler does not complain ...
    const MY_CONST: X = Bar;
}

fn main() {
    println!("{:?}", Foo::MY_CONST);
}
```
```
error[E0308]: mismatched types
  --> src/main.rs:13:25
   |
13 |     const MY_CONST: X = Bar;
   |                         ^^^ expected trait X, found struct `Bar`
   |
   = note: expected type `(dyn X + 'static)`
              found type `Bar`
```

The `X` is actually the same as `dyn X`, which is its own type, so an instance of that type is expected. Because of confusing stuff like this the plan was to remove the `Trait` syntax in favor of `dyn Trait`, but I'm not sure what exactly the plan is since the syntax evidently still works.

**Comment by jonas-schievink** [comment]
Closing since this does not seem a bug.
